fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/Compensation.java

```diff
@@ -302,6 +302,8 @@ static Compensation impossibleCompensation() {
     /**
      * Returns a specific compensation object that uses a mapping between predicates from a query to predicates that
      * are used for the application of the compensation.
+     * @param isImpossible an indicator if this compensation can be applied, if not, this compensation
```

Comment:
The changes in this file fix a problem that only got exposed by the other changes. Consider two matches and two predicates in a `SelectExpression`. Each match can sarg one predicate and need to be compensated for a remaining predicate. Let's further say that match `1` and match `2` sarg the respective other predicate and therefore need to compensate the respective other predicate. Further there are cases where compensations are impossible (e.g. `EXISTS()` compensations may need to reapplied). Let's say match `1` has such an impossible compensation (previously we just wouldn't match), it cannot be planned by itself _BUT_ the compensation can be intersected with other compensations which may eliminate the need for the impossible compensation to be planned at all.

Reply:
Addendum: The approach here is a little bit like complex numbers and real numbers. A match can lead to a compensation which is considered impossible (or not plannable). That's like a complex number with an imaginary part. Performing operations other than planning the compensation, such intersecting this compensation with other compensations may lead you back to a plannable compensation (a real number).